walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -36,50 +39,171 @@ public int compare(Event a, Event b) {
       };
 
   /**
-   * Returns a list of time periods in which the meeting, specified by request, could happen. If one
-   * or more time slots exists so that both mandatory and optional attendees can attend, it returns
-   * those time slots. Otherwise, it returns the time slots that fit just the mandatory attendees.
+   * Returns a list of time periods in which the meeting, specified by request, could happen. If no
+   * time exists for all optional and mandatory attendees, find the time slot(s) that allow
+   * mandatory attendees and the greatest possible number of optional attendees to attend.
    *
-   * @param eventsCollection the events we know about
+   * @param events the events to be considered
    * @param request information about the meeting, including attendees, optional attendees, and how
    *     long it needs to be
    */
-  public Collection<TimeRange> query(Collection<Event> eventsCollection, MeetingRequest request) {
-    Collection<TimeRange> withOptionalAttendees = getMeetingTimes(eventsCollection, request, true);
-
-    // Special case: if no mandatory attendees and optional attendees' schedules cannot fit in a
-    // meeting, no meeting times are possible.
-    return !withOptionalAttendees.isEmpty() || request.getAttendees().isEmpty()
-        ? withOptionalAttendees
-        : getMeetingTimes(eventsCollection, request, false);
+  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+    return getOptimalMeetingTimes(
+        getMandatoryAttendeesMeetingTimes(
+            events, request.getDuration(), new HashSet<String>(request.getAttendees())),
+        getChanges(
+            getOptionalAttendeesFreeTimes(
+                events,
+                new HashSet<String>(request.getOptionalAttendees()),
+                request.getDuration())),
+        request.getDuration());
+  }
+
+  /**
+   * Returns all meeting times that allow the most optional attendees to attend. These times must
+   * also fall in a good window and must be at least minTime long. Utilizes two pointers: one to
+   * goodWindows, one to changeLog.
+   *
+   * @param goodWindows all meeting times satisfying mandatory attendees
+   * @param changeLog a change log of the number of available optional attendees over time
+   * @param minTime the minimum time a meeting must last for
+   */
+  private Collection<TimeRange> getOptimalMeetingTimes(
+      ArrayList<TimeRange> goodWindows,
+      ArrayList<Map.Entry<Integer, Integer>> changeLog,
+      long minTime) {
+
+    ArrayList<TimeRange> optimalMeetingTimes = new ArrayList<TimeRange>();
+
+    // i is a pointer in changeLog, j is a poitner in goodWindows.
+    for (int j = 0, i = 0, currAttendance = 0, bestAttendance = 0, prevTime = 0;
+        i < changeLog.size();
+        ++i) {
+      // Need to back up j in case we missed a good window.
+      j = Math.max(0, j - 1);
```

Comment:
Should this be `i - 1`?

Reply:
No, the strategy is to have j and i be pointers in two different arrays so the complexity if O(changeLog.size() + goodWindows.size())